File: Java/LV3/네트워크.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <functional>
#include <iostream>
#include <queue>
#include <map>
using namespace std;

vector<vector<int>> adjList(201);
bool visited[200]={false,};

void bfs(int start){
    queue<int> q;
    visited[start]=true;
    q.push(start);
    
    while(!q.empty()){
        int cur = q.front();
        q.pop();
        
        for(int k : adjList[cur]){
            if(visited[k]!=true){
                q.push(k);
                visited[k]=true;
            }
        }
    }
}

void dfs(int start){
    visited[start]=true;
    
    for(int k : adjList[start]){
        if(visited[k]!=true){
            dfs(k);
        }
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    
    // 인접리스트 세팅
    for(int i=0;i<n;i++){
        for(int j=0;j<n;j++){
            if(computers[i][j]==1){
                if(i!=j){
                    adjList[i].push_back(j);
                    adjList[j].push_back(i);
                }
            }
        }
    }
    
    for(int i=0;i<n;i++){
        if(visited[i]==false){
            dfs(i);
            //bfs(i);
            answer++;
        }

    }
    return answer;
}

// 0: 1
// 1: 0,2
// 2: 1

File: Java/LV3/네트워크.java
import java.util.*;

class Solution {
    
    
    public boolean[] visited = new boolean[200];
    public ArrayList<ArrayList<Integer>> adjList = new ArrayList<ArrayList<Integer>>();
    
    public void bfs(int start){
        Queue<Integer> q = new LinkedList<Integer>();
        q.add(start);
        visited[start]=true;
        
        while(!q.isEmpty()){
            int cur = q.peek();
            q.poll();
            
            for(int k : adjList.get(cur)){
                if(visited[k]==false){
                    q.add(k);
                    visited[k]=true;
                }
            }
        }
        
    }
    
    public void dfs(int start){
        visited[start]=true;
        
        for(int k : adjList.get(start)){
            if(visited[k]==false){
                dfs(k);
        
            }
        }
    }
    public int solution(int n, int[][] computers) {
        int answer = 0;
        
        for(int i=0;i<n;i++){
            adjList.add(new ArrayList<Integer>());
        }
        
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(computers[i][j]==1){
                    if(i!=j){
                        adjList.get(i).add(j);
                        adjList.get(j).add(i);
                    }
                }
            }
        }
        
        for(int i=0;i<n;i++){
            if(visited[i]==false){
                bfs(i);
                //dfs(i);
                answer++;
            }
        }
        
        return answer;
    }
}
